crates/sui-package-management/src/lib.rs

```diff
@@ -73,3 +90,65 @@ pub async fn update_lock_file(
     lock.commit(lock_file)?;
     Ok(())
 }
+
+/// Find the published on-chain ID in the `Move.lock` or `Move.toml` file.
+/// A chain ID of `None` means that we will only try to resolve a published ID from the Move.toml.
+/// The published ID is resolved from the `Move.toml` if the Move.lock does not exist.
+/// Else, we resolve from the `Move.lock`, where addresses are automatically
+/// managed. If conflicting IDs are found in the `Move.lock` vs. `Move.toml`, a
+/// "Conflict" error message returns.
+pub fn resolve_published_id(
+    package: &Package,
+    chain_id: Option<String>,
+) -> Result<ObjectID, PublishedAtError> {
+    // Look up a valid `published-at` in the `Move.toml` first, which we'll
+    // return if the Move.lock does not manage addresses.
+    let published_id_in_manifest = match published_at_property(package) {
+        Ok(v) => Some(v),
+        Err(PublishedAtError::NotPresent) => None,
+        Err(e) => return Err(e), // An existing but invalid `published-at` in `Move.toml` should fail early.
+    };
+
+    let lock = package.package_path.join(SourcePackageLayout::Lock.path());
+    let mut lock_file = match File::open(lock.clone()) {
+        Ok(f) => f,
+        Err(_) => match published_id_in_manifest {
+            Some(v) => {
+                return ObjectID::from_str(v.as_str())
+                    .map_err(|_| PublishedAtError::Invalid(v.to_owned()))
+            }
+            None => return Err(PublishedAtError::NotPresent),
+        },
+    };
```

Comment:
nit (optional): `let ... else`?

```suggestion
    let Ok(mut lock_file) = File::open(lock.clone()) else {
        return match published_id_in_manifest {
            Some(v) => {
                return ObjectID::from_str(v.as_str())
                    .map_err(|_| PublishedAtError::Invalid(v.to_owned()))
            }
            None => return Err(PublishedAtError::NotPresent),
        },
    };
```

Reply:
Done!